modules/stanford_earth_capx/src/EarthCapxInfo.php

```diff
@@ -0,0 +1,249 @@
+   * if we do, return its fid. if we don't, return false
+   * @param array $source
+   */
+  public function currentProfilePhotoId($source = []) {
+    if (empty($source['profile_photo']) || empty($this->profilePhotoFid)
+      || empty($this->profilePhotoTimestamp)) {
+      return false;
+    }
+    $photo_ts = $this->getPhotoTimestamp($source['profile_photo']);
+    if ($photo_ts == $this->profilePhotoTimestamp) {
+      return $this->profilePhotoFid;
+    } else {
+      return false;
+    }
+  }
+  
+  /**
+   * return true if profile should be updated because it is new
+   * or because the etag has changed.
+   * @param array $source
+   */
+  public function getOkayToUpdateProfile($source = [], $photoId = 0) {
+    $oktoupdate = false;
+    $msg = new MigrateMessage();
+    if (empty($source['sunetid']) ||
+      $this->status == self::EARTH_CAPX_INFO_INVALID ||
+      $source['sunetid'] !== $this->sunetid) {
+        $msg->display('Unable to validate new profile information.', 'error');
+    } else if ($this->status == self::EARTH_CAPX_INFO_NEW) {
+      $oktoupdate = true;
+    } else if ($this->status == self::EARTH_CAPX_INFO_FOUND) {
+      $source_etag = '';
+      if (!empty($source['etag'])) {
+        $source_etag = $source['etag'];
+      }
+      if ($this->etag !== $source_etag || $this->profilePhotoFid !== $photoId) {
+        $oktoupdate = true;
+      }
+    }
+    return $oktoupdate;
+  }
+  
+  /**
+   * update the table with information from the source array and destination id
+   * only do the operation if information has changed
+   * 
+   * @param array $source
+   */
+  public function setInfoRecord($source = [], $entity_id = 0, $photo_id = 0) {
+    // see if we have valid sunetids
+    $msg = new MigrateMessage();
+    if (empty($source['sunetid']) ||
+      $this->status == self::EARTH_CAPX_INFO_INVALID) {
+        $msg->display('Unable to update EarthCapxInfo table. Missing source id.','error');
+        return;
+    }
+    if ($source['sunetid'] !== $this->sunetid) {
+      $msg->display('Unable to update EarthCapxInfo table. Mismatched ids: ' .
+        $source['sunetid'] . ', ' . $this->sunetid);
+      return;
+    }
+
+    // get the fields from the source array
+    $source_etag = '';
+    if (!empty($source['etag'])) $source_etag = $source['etag'];
+    
+    $source_ts = '';
+    if (!empty($source['profile_photo'])) {
+      $source_ts = $this->getPhotoTimestamp($source['profile_photo']);
+    }
+
+    $photo_id = intval($photo_id);
+    
+    // if existing in table, see if we need to update
+    if ($this->status == self::EARTH_CAPX_INFO_FOUND) {
+      if ($this->etag !== $source_etag ||
+        $this->profilePhotoTimestamp !== $source_ts ||
+        $this->profilePhotoFid !== $photo_id ||
+        $this->entityId != $entity_id) {
+          // the information is different, so delete record and set status = NEW
+          \Drupal::database()->delete(self::EARTH_CAPX_INFO_TABLE)
```

Comment:
Is this so the new status gets called? Is it dangerous to delete the record and insert new? What if the new insert fails? Does the entity get orphaned in Drupal?

Reply:
I found a number of articles claiming that it is more efficient to delete and insert records than it is to update them. If the insert should fail, the profile has still been imported and the next update will just re-import it. The migration system will still have a record of the import at this stage, and so the next update should reimport the profile even if nothing has changed. In any case, I have added exception handling around the insert command.

Reply:
Ok, this logging is good. It will help troubleshoot a profile if it constantly fails to write due to an invalid sunet format or some other unforseen issue.